pkgs/development/cuda-modules/setup-hooks/extension.nix

```diff
@@ -1,74 +1,74 @@
-final: _: {
+# NOTE(@connorbaker): None of the functions in this attribute set should need to access _final or _prev.
+# As such, they are prefixed with an underscore -- everything should be doable with the spliced package sets
+# provided to each function in the attribute set by `final.callPackage`.
+let
+  createSetupHooks =
+    setupHooksAttrs: final: prev:
+    let
+      # It is imperative that we use `final.callPackage` to create these setup hooks, as it allows us access to the spliced
+      # package sets.
+      inherit (final) callPackage;
+
+      # NOTE(@connorbaker): We MUST use `lib` from `prev` because the attribute names CAN NOT depend on `final`.
+      inherit (prev.lib.attrsets) mapAttrs;
+    in
+    mapAttrs (_: value: callPackage value { }) setupHooksAttrs;
+in
+createSetupHooks {
   # Internal hook, used by cudatoolkit and cuda redist packages
   # to accommodate automatic CUDAToolkit_ROOT construction
   markForCudatoolkitRootHook =
-    final.callPackage
-      (
-        {makeSetupHook}:
-        makeSetupHook {name = "mark-for-cudatoolkit-root-hook";} ./mark-for-cudatoolkit-root-hook.sh
-      )
-      {};
+    { makeSetupHook }:
+    makeSetupHook { name = "mark-for-cudatoolkit-root-hook"; } ./mark-for-cudatoolkit-root-hook.sh;
 
   # Currently propagated by cuda_nvcc or cudatoolkit, rather than used directly
   setupCudaHook =
-    (final.callPackage
-      (
-        {makeSetupHook, backendStdenv}:
-        makeSetupHook
-          {
-            name = "setup-cuda-hook";
-
-            substitutions.setupCudaHook = placeholder "out";
-
-            # Point NVCC at a compatible compiler
-            substitutions.ccRoot = "${backendStdenv.cc}";
+    { backendStdenv, makeSetupHook }:
+    makeSetupHook
+      {
+        name = "setup-cuda-hook";
 
-            # Required in addition to ccRoot as otherwise bin/gcc is looked up
-            # when building CMakeCUDACompilerId.cu
-            substitutions.ccFullPath = "${backendStdenv.cc}/bin/${backendStdenv.cc.targetPrefix}c++";
-          }
-          ./setup-cuda-hook.sh
-      )
-      {}
-    );
+        substitutions = {
+          # Required in addition to ccRoot as otherwise bin/gcc is looked up
+          # when building CMakeCUDACompilerId.cu
+          ccFullPath = "${backendStdenv.cc}/bin/${backendStdenv.cc.targetPrefix}c++";
+          # Point NVCC at a compatible compiler
+          ccRoot = "${backendStdenv.cc}";
+          setupCudaHook = placeholder "out";
+        };
```

Comment:
TODO @ConnorBaker

Are these correct? Do we need to use a spliced version? Unclear whether `makeSetupHook` does anything on the backend to draw these from `buildPackages` (`pkgsBuildHost`) or uses the default `pkgs` (`pkgsHostTarget`).

Reply:
We should use the current host's compiler. Then we take the hook from the BuildHost slice, so its "host" is really our "build"